bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -148,13 +177,70 @@ void ReplicaImp::messageHandler(MessageBase *msg) {
       }
     }
   }
-  if (!isCollectingState() && validateMessage(trueTypeObj)) {
+
```

Comment:
I think I'm confused about the above code. This is some code you had in the past specific to pre-preapre , right?
I was assuming you are planning to embed it in the common solution?
In that case, even the code handle preprepare async. , why is it left and not moved to the validation handler?

Reply:
There are 3 kinds of validation case and 2 kind of message handlers.
Validation case:
1) Generic sync validation.
2) Generic Async validation
3) PrePrepare specific Async validation.

When a message comes its handled by messageHandler() which decides which of the above validation will take place.
If its sync validation then the onMessage is also called.
If it's any kind of async validation, then the validated message handler will be called later. The validated message handler will have to check the pruning status and state transfer status again, before calling the onMessage.
With the above background, below is the ans to the ques:
1) This code is not for preprepare, its for all messages. This is for sync validation.
2) Currently I haven't merged the async validation, since the pre process result validation is specific to preprepare. This flow need to be completely checked, so it should be taken separately. I think its doable, and it will definitely reduce the complexity.

Reply:
@arc-vmware 
OK thanks for the explenation - answer 1 - I compeltely understood, for answer 2 - I now I understand you leave the case of merging the cases for the future.